File: core/geosite.py
"""Список доменов российских сервисов для обхода VPN (geosite-bypass).

Источник — v2fly/domain-list-community, файл "category-ru": кураторский
список (банки, госуслуги, ритейл, медиа, IT-компании и т.д.), устроенный
через рекурсивные "include:" на другие файлы того же репозитория — при
обновлении разворачиваем их в плоский список доменов.

Этот список нужен в дополнение к geoip (core/geoip.py): geoip исключает
по IP-диапазонам России, но сайты, которые физически хостятся на
зарубежном CDN (Cloudflare и т.п.), под него не попадают — для них и
нужен список доменов, передаваемый в наш патч hysteria2 (directDomains),
который сам узнаёт текущий IP через DNS-сниффинг.
"""
import concurrent.futures
from pathlib import Path

import requests

BUNDLED_DIR = Path(__file__).resolve().parent.parent / "data" / "geosite"
USER_DIR = Path.home() / ".config" / "vroxory-vpn" / "geosite"

SOURCE_BASE = "https://raw.githubusercontent.com/v2fly/domain-list-community/master/data/"
ROOT_CATEGORY = "category-ru"


def _parse_domain_file(path: Path) -> list:
    if not path.exists():
        return []
    with open(path, "r", encoding="utf-8") as f:
        return [line.strip() for line in f if line.strip() and not line.startswith("#")]


def get_ru_domains() -> list:
    """Возвращает список доменов. Файл из USER_DIR (после "Обновить")
    приоритетнее встроенного снимка из BUNDLED_DIR."""
    path = USER_DIR / "ru_domains.txt"
    if not path.exists():
        path = BUNDLED_DIR / "ru_domains.txt"
    return _parse_domain_file(path)


def last_updated() -> str:
    user_file = USER_DIR / "ru_domains.txt"
    if user_file.exists():
        import datetime
        return datetime.datetime.fromtimestamp(user_file.stat().st_mtime).strftime("%d.%m.%Y %H:%M")
    return "встроенный список (из установки)"


def current_size_kb() -> float:
    path = USER_DIR / "ru_domains.txt"
    if not path.exists():
        path = BUNDLED_DIR / "ru_domains.txt"
    return path.stat().st_size / 1024 if path.exists() else 0.0


def _fetch_file(name: str, timeout: int, retries: int = 2) -> str | None:
    for _ in range(retries):
        try:
            resp = requests.get(SOURCE_BASE + name, timeout=timeout)
            resp.raise_for_status()
            return resp.text
        except requests.RequestException:
            continue
    return None


def update_ru_domains(timeout: int = 15) -> dict:
    """Рекурсивно разворачивает category-ru (и все include: внутри) в
    плоский список доменов и сохраняет в USER_DIR. Возвращает
    {"count": кол-во доменов, "bytes": размер сохранённого файла}.
    Файлы, которые не удалось скачать (например, переименованные в
    апстриме), просто пропускаются — список всё равно получается
    рабочим, просто чуть менее полным."""
    seen_files = set()
    domains = set()
    pending = {ROOT_CATEGORY}

    with concurrent.futures.ThreadPoolExecutor(max_workers=24) as pool:
        while pending:
            to_fetch = [n for n in pending if n not in seen_files]
            seen_files.update(to_fetch)
            pending = set()
            if not to_fetch:
                break
            fetched = pool.map(lambda name: (name, _fetch_file(name, timeout)), to_fetch)
            for name, text in fetched:
                if text is None:
                    continue
                for line in text.splitlines():
                    line = line.strip()
                    if not line or line.startswith("#"):
                        continue
                    line = line.split("#")[0].strip()
                    if not line:
                        continue
                    if line.startswith("include:"):
                        inc = line[len("include:"):].strip()
                        if inc not in seen_files:
                            pending.add(inc)
                        continue
                    entry = line.split("@")[0].strip()
                    for prefix in ("full:", "domain:"):
                        if entry.startswith(prefix):
                            entry = entry[len(prefix):]
                            break
                    if entry.startswith(("regexp:", "keyword:")) or not entry:
                        continue
                    domains.add(entry.lower())

    if not domains:
        raise ValueError("Не удалось скачать ни одного домена из category-ru")

    USER_DIR.mkdir(parents=True, exist_ok=True)
    text = "\n".join(sorted(domains)) + "\n"
    path = USER_DIR / "ru_domains.txt"
    with open(path, "w", encoding="utf-8") as f:
        f.write(text)

    return {"count": len(domains), "bytes": len(text.encode("utf-8"))}

File: app/src-tauri/src/geosite.rs
//! Список доменов российских сервисов для обхода VPN (geosite-bypass) —
//! порт core/geosite.py (ветка main). Источник —
//! v2fly/domain-list-community, category-ru с рекурсивными include:.

use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

use serde::Serialize;
use tauri::AppHandle;

use crate::resources;

const SOURCE_BASE: &str = "https://raw.githubusercontent.com/v2fly/domain-list-community/master/data/";
const ROOT_CATEGORY: &str = "category-ru";

fn user_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(".config/vroxory-vpn/geosite")
}

fn parse_domain_file(path: &Path) -> Vec<String> {
    fs::read_to_string(path)
        .map(|content| {
            content
                .lines()
                .map(str::trim)
                .filter(|l| !l.is_empty() && !l.starts_with('#'))
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default()
}

pub fn get_ru_domains(app: &AppHandle) -> Result<Vec<String>, String> {
    let user_file = user_dir().join("ru_domains.txt");
    let path = if user_file.exists() {
        user_file
    } else {
        resources::resolve(app, "resources/geosite/ru_domains.txt")?
    };
    Ok(parse_domain_file(&path))
}

#[derive(Serialize)]
pub struct UpdateResult {
    pub count: usize,
    pub bytes: usize,
}

async fn fetch_file(client: &reqwest::Client, name: &str, retries: u32) -> Option<String> {
    let url = format!("{SOURCE_BASE}{name}");
    for _ in 0..retries {
        let Ok(resp) = client
            .get(&url)
            .timeout(std::time::Duration::from_secs(15))
            .send()
            .await
        else {
            continue;
        };
        let Ok(resp) = resp.error_for_status() else {
            continue;
        };
        if let Ok(text) = resp.text().await {
            return Some(text);
        }
    }
    None
}

pub async fn update_ru_domains() -> Result<UpdateResult, String> {
    let client = reqwest::Client::new();
    let mut seen_files: HashSet<String> = HashSet::new();
    let mut domains: HashSet<String> = HashSet::new();
    let mut pending: HashSet<String> = HashSet::from([ROOT_CATEGORY.to_string()]);

    while !pending.is_empty() {
        let to_fetch: Vec<String> = pending
            .iter()
            .filter(|n| !seen_files.contains(*n))
            .cloned()
            .collect();
        seen_files.extend(to_fetch.iter().cloned());
        pending.clear();
        if to_fetch.is_empty() {
            break;
        }

        let fetches = to_fetch.iter().map(|name| {
            let client = client.clone();
            let name = name.clone();
            async move {
                let text = fetch_file(&client, &name, 2).await;
                (name, text)
            }
        });
        let results = futures::future::join_all(fetches).await;

        for (_name, text) in results {
            let Some(text) = text else { continue };
            for raw_line in text.lines() {
                let line = raw_line.trim();
                if line.is_empty() || line.starts_with('#') {
                    continue;
                }
                let line = line.split('#').next().unwrap_or("").trim();
                if line.is_empty() {
                    continue;
                }
                if let Some(inc) = line.strip_prefix("include:") {
                    let inc = inc.trim().to_string();
                    if !seen_files.contains(&inc) {
                        pending.insert(inc);
                    }
                    continue;
                }
                let mut entry = line.split('@').next().unwrap_or("").trim();
                for prefix in ["full:", "domain:"] {
                    if let Some(stripped) = entry.strip_prefix(prefix) {
                        entry = stripped;
                        break;
                    }
                }
                if entry.is_empty() || entry.starts_with("regexp:") || entry.starts_with("keyword:") {
                    continue;
                }
                domains.insert(entry.to_lowercase());
            }
        }
    }

    if domains.is_empty() {
        return Err("Не удалось скачать ни одного домена из category-ru".into());
    }

    let dir = user_dir();
    fs::create_dir_all(&dir).map_err(|e| e.to_string())?;
    let mut sorted: Vec<String> = domains.into_iter().collect();
    sorted.sort();
    let text = sorted.join("\n") + "\n";
    fs::write(dir.join("ru_domains.txt"), &text).map_err(|e| e.to_string())?;

    Ok(UpdateResult {
        count: sorted.len(),
        bytes: text.len(),
    })
}
